src/v/rpc/parse_utils.h

```diff
@@ -235,11 +235,22 @@ struct default_message_codec {
  *
  * example:
  *   using echo_service_v0 = echo_service_base<v0_message_codec>;
+ *
+ * Note that for serde-supported messages a vassert(false) is generated. First,
+ * the v0_message_encoder is only used in tests. Second, serde usage is not
+ * possible in v0 servers, so this restriction is realistic. And from a
+ * practical standpoint this allows us to avoid bifurcation of services (or more
+ * sfinae magic) in tests so that serde-only types were never present within a
```

Comment:
I think the sfinae is simple with concepts, but I'd have to compile it. I'll come back if there's time.